File: Week 10/2.SpiralAndDiagonalTraversal.java
/*
 * WEEK 10 - 2D ARRAYS & MATRIX
 * Topic: Spiral Traversal & Diagonal Traversal
 *
 * SPIRAL TRAVERSAL:
 * Traverse matrix in spiral order: right → down → left → up → repeat.
 * Use four boundaries: top, bottom, left, right.
 * Shrink boundaries as you go.
 * Time: O(m*n), Space: O(1) extra (or O(m*n) for result list)
 *
 * DIAGONAL TRAVERSAL:
 * Print all diagonals of a matrix (anti-diagonals: direction toggles).
 * Time: O(m*n), Space: O(1)
 *
 * SEARCH IN ROW-AND-COLUMN SORTED MATRIX:
 * Matrix where each row and column is sorted.
 * Start at top-right corner:
 * - If element == target: found!
 * - If element > target: move left (eliminate column)
 * - If element < target: move down (eliminate row)
 * Time: O(m + n), much better than O(m*n) brute force
 */

import java.util.ArrayList;
import java.util.List;

public class SpiralAndDiagonalTraversal {

    // Spiral order traversal
    static List<Integer> spiralOrder(int[][] mat) {
        List<Integer> result = new ArrayList<>();
        if (mat == null || mat.length == 0) return result;
        int top = 0, bottom = mat.length - 1;
        int left = 0, right = mat[0].length - 1;

        while (top <= bottom && left <= right) {
            // Traverse right
            for (int i = left; i <= right; i++) result.add(mat[top][i]);
            top++;
            // Traverse down
            for (int i = top; i <= bottom; i++) result.add(mat[i][right]);
            right--;
            // Traverse left (if still valid)
            if (top <= bottom) {
                for (int i = right; i >= left; i--) result.add(mat[bottom][i]);
                bottom--;
            }
            // Traverse up (if still valid)
            if (left <= right) {
                for (int i = bottom; i >= top; i--) result.add(mat[i][left]);
                left++;
            }
        }
        return result;
    }

    // Diagonal (anti-diagonal zigzag) traversal
    static int[] diagonalOrder(int[][] mat) {
        int m = mat.length, n = mat[0].length;
        int[] result = new int[m * n];
        int idx = 0;
        for (int d = 0; d < m + n - 1; d++) {
            if (d % 2 == 0) { // going up
                int r = Math.min(d, m - 1);
                int c = d - r;
                while (r >= 0 && c < n) result[idx++] = mat[r--][c++];
            } else { // going down
                int c = Math.min(d, n - 1);
                int r = d - c;
                while (c >= 0 && r < m) result[idx++] = mat[r++][c--];
            }
        }
        return result;
    }

    // Search in row-and-column sorted matrix
    static boolean searchSortedMatrix(int[][] mat, int target) {
        int row = 0, col = mat[0].length - 1; // start at top-right
        while (row < mat.length && col >= 0) {
            if (mat[row][col] == target) return true;
            else if (mat[row][col] > target) col--; // too big, go left
            else row++;                              // too small, go down
        }
        return false;
    }

    public static void main(String[] args) {
        int[][] mat = {
            {1,  2,  3,  4},
            {5,  6,  7,  8},
            {9, 10, 11, 12}
        };

        System.out.println("Matrix:");
        for (int[] row : mat) System.out.println(java.util.Arrays.toString(row));

        System.out.println("\nSpiral order: " + spiralOrder(mat));
        // Expected: [1,2,3,4,8,12,11,10,9,5,6,7]

        int[][] sq = {{1,2,3},{4,5,6},{7,8,9}};
        System.out.println("\nSquare matrix:");
        for (int[] row : sq) System.out.println(java.util.Arrays.toString(row));
        System.out.println("Spiral: " + spiralOrder(sq));
        System.out.println("Diagonal: " + java.util.Arrays.toString(diagonalOrder(sq)));
        // Expected diagonal: [1,2,4,7,5,3,6,8,9]

        // Search in sorted matrix
        int[][] sorted = {
            { 1,  4,  7, 11, 15},
            { 2,  5,  8, 12, 19},
            { 3,  6,  9, 16, 22},
            {10, 13, 14, 17, 24},
            {18, 21, 23, 26, 30}
        };
        System.out.println("\nSearch in sorted matrix:");
        System.out.println("Search 5:  " + searchSortedMatrix(sorted, 5));   // true
        System.out.println("Search 20: " + searchSortedMatrix(sorted, 20));  // false
        System.out.println("Search 1:  " + searchSortedMatrix(sorted, 1));   // true
        System.out.println("Search 30: " + searchSortedMatrix(sorted, 30));  // true
    }
}

File: Week 10/python/2.spiral_and_diagonal_traversal.py
"""
WEEK 10 - PYTHON DSA
Topic: Spiral & Diagonal Traversal + Sorted-Matrix Search
File: 2.spiral_and_diagonal_traversal.py

CONCEPT:
    Spiral traversal walks the perimeter inward: right -> down -> left -> up,
    shrinking boundaries each lap.
    Diagonal (anti-diagonal zigzag) traversal alternates direction on each
    diagonal index.
    Row/column-sorted-matrix search starts at the top-right and moves
    left/down to eliminate a row or column per step.

KEY POINTS:
    - Spiral: maintain four boundaries (top, bottom, left, right).
    - Diagonal: iterate d from 0 to m+n-2; even d goes up-right, odd goes down-left.
    - Sorted-matrix search: O(m + n) instead of O(m*n) brute force.

ALGORITHM / APPROACH:
    Spiral:
        while top <= bottom and left <= right:
            traverse top row    [left..right];  top++
            traverse right col  [top..bottom];  right--
            if top <= bottom: traverse bottom row [right..left]; bottom--
            if left <= right: traverse left col   [bottom..top]; left++

    Diagonal:
        for d in 0..m+n-2:
            if d even: start (min(d,m-1), d - min(d,m-1)); walk r--, c++
            else:      start (d - min(d,n-1), min(d,n-1)); walk r++, c--

    Sorted-matrix search:
        start at top-right; if equal -> found; if > target -> col--; else row++

PYTHON-SPECIFIC NOTES:
    - Tuple unpacking and ranges (range(start, stop, -1)) keep traversal concise.
    - any() / all() not directly used here, but lists naturally hold the result.

DRY RUN:
    Spiral on:
        1 2 3 4
        5 6 7 8
        9 10 11 12
    => [1,2,3,4,8,12,11,10,9,5,6,7]

    Diagonal zigzag on:
        1 2 3
        4 5 6
        7 8 9
    => [1, 2,4, 7,5,3, 6,8, 9]

COMPLEXITY:
    Spiral / Diagonal: O(m*n) time, O(m*n) for output.
    Sorted search    : O(m + n) time, O(1) space.
"""

from typing import List

Matrix = List[List[int]]


def spiral_order(mat: Matrix) -> List[int]:
    result: List[int] = []
    if not mat or not mat[0]:
        return result
    top, bottom = 0, len(mat) - 1
    left, right = 0, len(mat[0]) - 1
    while top <= bottom and left <= right:
        # right along top row
        for i in range(left, right + 1):
            result.append(mat[top][i])
        top += 1
        # down along right column
        for i in range(top, bottom + 1):
            result.append(mat[i][right])
        right -= 1
        # left along bottom row
        if top <= bottom:
            for i in range(right, left - 1, -1):
                result.append(mat[bottom][i])
            bottom -= 1
        # up along left column
        if left <= right:
            for i in range(bottom, top - 1, -1):
                result.append(mat[i][left])
            left += 1
    return result


def diagonal_order(mat: Matrix) -> List[int]:
    """Anti-diagonal zigzag traversal."""
    m, n = len(mat), len(mat[0])
    result: List[int] = []
    for d in range(m + n - 1):
        if d % 2 == 0:  # going up
            r = min(d, m - 1)
            c = d - r
            while r >= 0 and c < n:
                result.append(mat[r][c])
                r -= 1
                c += 1
        else:  # going down
            c = min(d, n - 1)
            r = d - c
            while c >= 0 and r < m:
                result.append(mat[r][c])
                r += 1
                c -= 1
    return result


def search_sorted_matrix(mat: Matrix, target: int) -> bool:
    """Row- and column-sorted matrix search starting from top-right."""
    if not mat or not mat[0]:
        return False
    row, col = 0, len(mat[0]) - 1
    while row < len(mat) and col >= 0:
        if mat[row][col] == target:
            return True
        elif mat[row][col] > target:
            col -= 1
        else:
            row += 1
    return False


def main() -> None:
    mat = [
        [1, 2, 3, 4],
        [5, 6, 7, 8],
        [9, 10, 11, 12],
    ]
    print("Matrix:")
    for row in mat:
        print(row)
    print(f"\nSpiral order: {spiral_order(mat)}")
    # Expected: [1,2,3,4,8,12,11,10,9,5,6,7]

    sq = [[1, 2, 3], [4, 5, 6], [7, 8, 9]]
    print("\nSquare matrix:")
    for row in sq:
        print(row)
    print(f"Spiral:   {spiral_order(sq)}")
    print(f"Diagonal: {diagonal_order(sq)}")
    # Expected diagonal: [1,2,4,7,5,3,6,8,9]

    sorted_mat = [
        [1, 4, 7, 11, 15],
        [2, 5, 8, 12, 19],
        [3, 6, 9, 16, 22],
        [10, 13, 14, 17, 24],
        [18, 21, 23, 26, 30],
    ]
    print("\nSearch in sorted matrix:")
    print(f"Search 5:  {search_sorted_matrix(sorted_mat, 5)}")   # True
    print(f"Search 20: {search_sorted_matrix(sorted_mat, 20)}")  # False
    print(f"Search 1:  {search_sorted_matrix(sorted_mat, 1)}")   # True
    print(f"Search 30: {search_sorted_matrix(sorted_mat, 30)}")  # True


if __name__ == "__main__":
    main()


"""
NOTES — Python vs Java:
    - range(start, stop, -1) handles descending loops cleanly.
    - List append + list literals replace Java ArrayList/Arrays.
    - The four-boundary spiral idiom is identical across languages.
"""
